hooks.js example causes a build error

**Describe the bug**

When the following `hooks.js` file exists, `npm run build` throws an error. (However `npm run dev` works without error and the site runs perfectly.)

**To Reproduce**

1. Add `src/hooks.js`

```js
export async function handle(request, render) {
  const response = await render(request);

  return {
    ...response,
    headers: {
      ...response.headers,
      'x-custom-header': 'potato'
    }
  };
}
```

2. Run `npm run build`.

The command will throw this error:

```
✓ 38 modules transformed.
.svelte/output/server/app.js      193.07kb
.svelte/output/server/style.css   7.76kb

Run npm start to try your app locally.

> Using @sveltejs/adapter-node
TypeError: Cannot read property 'headers' of undefined
    at Object.handle (file:///Users/me/sveltekit-demo/.svelte/output/server/app.js:1890:19)
    at async ssr (file:///Users/me/sveltekit-demo/.svelte/output/server/app.js:1677:12)
    at async visit (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:531:20)
    at async prerender (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:682:5)
    at async Builder.prerender (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:743:4)
    at async adapt (/Users/me/sveltekit-demo/node_modules/@sveltejs/adapter-node/index.js:26:4)
    at async adapt (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:765:2)
    at async file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/cli.js:598:5
> 500 /about
Error: 500 /about
    at error (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:523:11)
    at visit (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:574:5)
    at async prerender (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:682:5)
    at async Builder.prerender (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:743:4)
    at async adapt (/Users/me/sveltekit-demo/node_modules/@sveltejs/adapter-node/index.js:26:4)
    at async adapt (file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/chunks/index5.js:765:2)
    at async file:///Users/me/sveltekit-demo/node_modules/@sveltejs/kit/dist/cli.js:598:5
npm ERR! code 1
npm ERR! path /Users/me/sveltekit-demo
npm ERR! command failed
npm ERR! command sh -c svelte-kit build
```

`about.svelte` is nothing special:
```
<h1>About this site</h1>
<p>TODO...</p>
```

`svelte.config.cjs` contains adapter as: `adapter: node(),`.

**Information about your SvelteKit Installation:**

```
  System:
    OS: macOS 11.2.2
    CPU: (8) arm64 Apple M1
    Shell: 5.8 - /bin/zsh
 npmPackages:
    @sveltejs/kit: next => 1.0.0-next.64 
    svelte: ^3.35.0 => 3.35.0 
```

**Severity**

Minor

--- a/.changeset/quiet-mangos-shop.md
+++ b/.changeset/quiet-mangos-shop.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+Always return a response from render function in handle

--- a/documentation/docs/04-hooks.md
+++ b/documentation/docs/04-hooks.md
@@ -97,7 +97,7 @@ type Response = {
 
 type Handle<Context = any> = (
 	request: Request<Context>,
-	render: (request: Request<Context>) => Response | Promise<Response>
+	render: (request: Request<Context>) => Promise<Response>
 ) => Response | Promise<Response>;
 ```
 

--- a/packages/kit/src/core/adapt/prerender.js
+++ b/packages/kit/src/core/adapt/prerender.js
@@ -123,10 +123,10 @@ export async function prerender({ cwd, out, log, config, force }) {
 				return;
 			}
 
-			if (response_type === OK) {
+			if (rendered.status === 200) {
 				log.info(`${rendered.status} ${path}`);
 				writeFileSync(file, rendered.body); // TODO minify where possible?
-			} else {
+			} else if (response_type !== OK) {
 				error(rendered.status, path);
 			}
 

--- a/packages/kit/src/core/adapt/prerender.js
+++ b/packages/kit/src/core/adapt/prerender.js
@@ -84,6 +84,9 @@ export async function prerender({ cwd, out, log, config, force }) {
 		if (seen.has(path)) return;
 		seen.add(path);
 
+		/** @type {Map<string, import('types').Response>} */
+		const dependencies = new Map();
+
 		const rendered = await app.render(
 			{
 				host: config.kit.host,
@@ -95,6 +98,7 @@ export async function prerender({ cwd, out, log, config, force }) {
 			},
 			{
 				local: true,
+				dependencies,
 				only_render_prerenderable_pages: !force,
 				get_static_file: (file) => readFileSync(join(config.kit.files.assets, file))
 			}
@@ -130,32 +134,27 @@ export async function prerender({ cwd, out, log, config, force }) {
 				error(rendered.status, path);
 			}
 
-			const { dependencies } = rendered;
+			dependencies.forEach((result, path) => {
+				const response_type = Math.floor(result.status / 100);
 
-			if (dependencies) {
-				for (const path in dependencies) {
-					const result = dependencies[path];
-					const response_type = Math.floor(result.status / 100);
+				const is_html = result.headers['content-type'] === 'text/html';
 
-					const is_html = result.headers['content-type'] === 'text/html';
-
-					const parts = path.split('/');
-					if (is_html && parts[parts.length - 1] !== 'index.html') {
-						parts.push('index.html');
-					}
-
-					const file = `${out}${parts.join('/')}`;
-					mkdirp(dirname(file));
-
-					writeFileSync(file, result.body);
-
-					if (response_type === OK) {
-						log.info(`${result.status} ${path}`);
-					} else {
-						error(result.status, path);
-					}
+				const parts = path.split('/');
+				if (is_html && parts[parts.length - 1] !== 'index.html') {
+					parts.push('index.html');
 				}
-			}
+
+				const file = `${out}${parts.join('/')}`;
+				mkdirp(dirname(file));
+
+				writeFileSync(file, result.body);
+
+				if (response_type === OK) {
+					log.info(`${result.status} ${path}`);
+				} else {
+					error(result.status, path);
+				}
+			});
 
 			if (is_html && config.kit.prerender.crawl) {
 				const cleaned = clean_html(rendered.body);

--- a/packages/kit/src/core/build/index.js
+++ b/packages/kit/src/core/build/index.js
@@ -354,6 +354,7 @@ async function build_server(
 			export function render(request, {
 				paths = ${s(config.kit.paths)},
 				local = false,
+				dependencies,
 				only_render_prerenderable_pages = false,
 				get_static_file
 			} = {}) {
@@ -371,6 +372,7 @@ async function build_server(
 					hooks,
 					dev: false,
 					amp: ${config.kit.amp},
+					dependencies,
 					only_render_prerenderable_pages,
 					app_dir: ${s(config.kit.appDir)},
 					get_component_path: id => ${s(`${config.kit.paths.assets}/${config.kit.appDir}/`)} + client_component_lookup[id],

--- a/packages/kit/src/runtime/server/endpoint.js
+++ b/packages/kit/src/runtime/server/endpoint.js
@@ -1,7 +1,7 @@
 /**
  * @param {import('types').Request} request
  * @param {import('types.internal').SSREndpoint} route
- * @returns {Promise<import('types.internal').ResponseWithDependencies>}
+ * @returns {Promise<import('types').Response>}
  */
 export default async function render_route(request, route) {
 	const mod = await route.load();

--- a/packages/kit/src/runtime/server/page.js
+++ b/packages/kit/src/runtime/server/page.js
@@ -240,11 +240,23 @@ async function get_response({ request, options, $session, route, status = 200, e
 	};
 
 	if (options.only_render_prerenderable_pages) {
-		if (error) return; // don't prerender an error page
+		if (error) {
+			return {
+				status,
+				headers: {},
+				body: error.message
+			};
+		}
 
 		// if the page has `export const prerender = true`, continue,
 		// otherwise bail out at this point
-		if (!page_component.prerender) return;
+		if (!page_component.prerender) {
+			return {
+				status: 204,
+				headers: {},
+				body: null
+			};
+		}
 	}
 
 	/** @type {{ head: string, html: string, css: string }} */

--- a/packages/kit/src/runtime/server/page.js
+++ b/packages/kit/src/runtime/server/page.js
@@ -16,12 +16,9 @@ const s = JSON.stringify;
  *   status: number;
  *   error: Error
  * }} opts
- * @returns {Promise<import('types.internal').ResponseWithDependencies>}
+ * @returns {Promise<import('types').Response>}
  */
 async function get_response({ request, options, $session, route, status = 200, error }) {
-	/** @type {Record<string, import('types.internal').ResponseWithDependencies>} */
-	const dependencies = {};
-
 	const serialized_session = try_serialize($session, (error) => {
 		throw new Error(`Failed to serialize session data: ${error.message}`);
 	});
@@ -146,9 +143,9 @@ async function get_response({ request, options, $session, route, status = 200, e
 				);
 
 				if (rendered) {
-					// TODO this is primarily for the benefit of the static case,
-					// but could it be used elsewhere?
-					dependencies[resolved] = rendered;
+					if (options.dependencies) {
+						options.dependencies.set(resolved, rendered);
+					}
 
 					response = new Response(rendered.body, {
 						status: rendered.status,
@@ -480,8 +477,7 @@ async function get_response({ request, options, $session, route, status = 200, e
 	return {
 		status,
 		headers,
-		body: options.template({ head, body }),
-		dependencies
+		body: options.template({ head, body })
 	};
 }
 
@@ -489,7 +485,7 @@ async function get_response({ request, options, $session, route, status = 200, e
  * @param {import('types').Request} request
  * @param {import('types.internal').SSRPage} route
  * @param {import('types.internal').SSRRenderOptions} options
- * @returns {Promise<import('types.internal').ResponseWithDependencies>}
+ * @returns {Promise<import('types').Response>}
  */
 export default async function render_page(request, route, options) {
 	if (options.initiator === route) {

--- a/packages/kit/types.internal.d.ts
+++ b/packages/kit/types.internal.d.ts
@@ -65,7 +65,7 @@ export type App = {
 		};
 		prerendering: boolean;
 	}) => void;
-	render: (incoming: Incoming, options: SSRRenderOptions) => ResponseWithDependencies;
+	render: (incoming: Incoming, options: SSRRenderOptions) => Response;
 };
 
 // TODO we want to differentiate between request headers, which
@@ -73,10 +73,6 @@ export type App = {
 // 'set-cookie' is a `string[]` (or at least `string | string[]`)
 // but this can't happen until TypeScript 4.3
 export type Headers = Record<string, string>;
-
-export type ResponseWithDependencies = Response & {
-	dependencies?: Record<string, Response>;
-};
 
 export type Page = {
 	host: string;
@@ -185,6 +181,7 @@ export type SSRRenderOptions = {
 	hooks?: Hooks;
 	dev?: boolean;
 	amp?: boolean;
+	dependencies?: Map<string, Response>;
 	only_render_prerenderable_pages?: boolean;
 	app_dir?: string;
 	get_component_path?: (id: string) => string;

